src/adb.js

```diff
@@ -487,6 +525,134 @@ class Adb {
+        // Get file size
+        _this.getFileSize(srcfile).then(stdout=>{
+           _this.log("Returned FileSize is "+stdout+ " Ko"); 
+           _this.fileSize = stdout;
+           // Creating pipe
+           _this.shell("mkfifo /backup.pipe").then((stdout) => {
+              _this.log("Pipe created !");
+              var lastSize = 0;
+              var progressInterval = setInterval(() => {
+                 _this.execCommandCustom(["stat", "-t", destfile]).then(stat => {
+                     progress ((lastSize / _this.fileSize)*100);
+                     lastSize = eval(stat.split(" ")[1])/1024;
+                     //_this.log("pushing: " + lastSize+"/"+_this.fileSize);
+                 }).catch(e => { clearInterval(progressInterval);_this.log("failed to stat: " + e);
+               });
+              }, 1000);
+
+              // Start the backup
+              _this.log("Starting Backup...");
+              _this.execCommand(["exec-out 'tar -cvp ",srcfile," 2>/backup.pipe' | dd of="+destfile,
+                                 " & ",(adbpath+"/adb -P"),_this.port,
+                                 " shell cat /backup.pipe", process.platform == "win32" ? ' | findstr /v "%]"' : ' | grep -v "%]"']).then( (error, stdout, stderr) => {
+	         _this.log("Backup Ended");
+                 clearInterval(progressInterval); 
+                 _this.shell("rm /backup.pipe").then( ()=> {
+                     _this.log("Pipe released.");
+                     resolve(); // Everything's Fine !
+                 }).catch(e=>{utils.log.warn; reject(e+" Unable to delete the pipe "+e.length);}); // Pipe Deletion
+              }).catch(e=>{utils.log.warn; reject(e+" Unable to backuping the device "+e.length);}); // Backup start
+           }).catch(e=>{utils.log.warn; reject(e+" Pipe creation failed "+e.length);}); // Pipe Creation
+        }).catch(e=>{utils.log.warn; reject(e+" Unable to get the partition's filesize "+e.length);}); // Get file size
+   });
+ }
+
+
+// AW : Restore file "srcfile" from the computer to "destfile" on the phone
+// TO-DO : This function should be moved to devices.js
+ restore(srcfile,destfile, adbpath, progress) {
+    var _this = this;
+    _this.log("Restore Function Entry, will restore "+srcfile+" on "+destfile+" adbpath:"+adbpath);
+    return new Promise(function(resolve, reject) {
+        _this.fileSize = 9999999999999; // Init to file size to a very high value while waiting for the real size.
+        // Get file size
+        _this.getFileSize(srcfile).then(stdout=>{
+           _this.log("Returned FileSize is "+stdout+ " Ko"); 
+           _this.fileSize = stdout;
+           // Creating pipe
+           _this.shell("mkfifo /restore.pipe").then((stdout) => {
+              _this.log("Pipe created !");
+              var lastSize = 0;
+              var progressInterval = setInterval(() => {
+                 _this.execCommandCustom(["stat", "-t", destfile]).then(stat => {
```

Comment:
Use [the fs module](https://nodejs.org/api/fs.html#fs_fs_stat_path_options_callback) for this instead.

Reply:
That's a really good idea !, it's done.